### ISSUE TITLE
memory leak found

### ISSUE THREAD

**Issue by 751620780** [Velocidex/go-pe#25]
When I run the following code, I find that the memory consumption is about 800 MB, but the memory consumption of the process commenting out ``pe.Imports()`` is very small.
```go
func PeParser(){
	pe, err := pe.NewPEFile(bytes.NewReader(d))
	if err != nil {
		need = true
		return
	}
	pe.Imports()
}
func main(){
        filepath.Walk(`pe_dir`, func(path string, info os.FileInfo, err error) error {
		if info.IsDir() {
			return nil
		}
               	data,err:=ioutil.ReadFile(path)
		if err!=nil{
			return err
		}
		PeParser(data)
	        return nil
	})
}
```

**Comment by 751620780** [comment]
I used the latest code and found that the problem was solved.

**Comment by scudette** [comment]
Can you find a specific pe file that triggers it? If you can privately share it with us (we are on discord https://docs.velociraptor.app/discord/ ) it would make it easier to fix.

**Comment by 751620780** [comment]
I used the latest code, tested 30000 PE samples, and found that there may still be a memory leak problem. I found that after testing 30000 samples, the memory occupied about 1800mb. Then the program executed ``time. Sleep (time. Second * 1000)`` and found that the memory decreased by 3MB per second during this period, and finally decreased to about 100MB. However, when I used the old version code, I found that the same 30000 PE files occupied about 13000mb of memory after parsing, and then decreased 3MB per second until the end of the program.
The code given in the example is not all. The code I call like this:
```go
var peSecNameKeys = map[*regexp.Regexp]bool{
	regexp.MustCompile(`^/\d+`):       false,
	regexp.MustCompile(`^\.textbss^`): false,
	regexp.MustCompile(`^\.idata\s+`): true,
	regexp.MustCompile(`^\s+`):        true,
	regexp.MustCompile(`^\.CRT$`):     false,
	regexp.MustCompile(`^\.bss$`):     false,
}

func PeFileter(d []byte) (need bool, reason string) {
	need = false
	defer func() {
		if e := recover(); e != nil {
			buf := make([]byte, 8192)
			runtime.Stack(buf, false)
		}
	}()
	pe, err := pe.NewPEFile(bytes.NewReader(d))
	if err != nil {
		need = true
		return
	}
	if len(pe.Sections) == 1 {
		need = true
		reason = `section count=1`
		return
	}
	if len(pe.Sections) > 10 {
		need = false
		reason = `section count>10`
		return
	}
	for _, sec := range pe.Sections {
		for re, filted := range peSecNameKeys {
			if re.MatchString(sec.Name) {
				need = filted
				reason = fmt.Sprintf(`found section name:"%s"`, sec.Name)
				return
			}
		}
	}
	pe.VersionInformation()
	pe.Imports()
	need = true
	reason = "nothing found"
	return
}
```
The 30000 samples used in the test were randomly downloaded from the network.

**Comment by scudette** [comment]
Can you please get your program to write the heap profile (example code here https://pkg.go.dev/runtime/pprof ). You can also use `import _ "net/http/pprof"` to bring up a web server on http://127.0.0.1:6060/debug/pprof/  that you can use to write heap profile along with time. It will immediately show who is allocating all the memory. 

If the memory decreases with time it might just be the garbage collector reclaiming memory over time and quite a normal thing. You can force gc collection with `runtime.GC()` which might be worth doing periodically

**Comment by 751620780** [comment]
Here is my test code：
```go
package main

import (
	"bytes"
	"fmt"
	_ "net/http/pprof"
	"os"
	"path/filepath"
	"runtime"

	"github.com/Velocidex/go-pe"
)

var data []byte = make([]byte, 50*1024*1024)

func PeFileter(d []byte) {

	pe, err := pe.NewPEFile(bytes.NewReader(d))
	if err != nil {
		return
	}
	pe.VersionInformation()
	pe.Imports()
}

func walkFun(path string, info os.FileInfo, err error) error {
	if err != nil {
		return err
	}
	if info.IsDir() {
		return nil
	}
	size := int(info.Size())
	if size > 50*1024*1024 {
		return nil
	}
	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()

	n, err := f.Read(data)
	if n != size {
		fmt.Printf("%s read failed, pass\n", path)
		return nil
	}
	if err != nil {
		return nil
	}
	PeFileter(data[:size])
	return nil
}

func MemoryInfo() uint64 {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	return m.Sys
}

func main() {
	filepath.Walk(`D:\samples\badpe`, walkFun)
	fmt.Printf("memory size=%.2f MB \n", float64(MemoryInfo())/1024.0/1024.0)
}
```
The program output is shown below:
```bash
PS R:\> go run .\main.go 
memory size=2048.45 MB
```
After testing, it is found that these three samples lead to memory leakage, which can be downloaded by [click here](https://github.com/Velocidex/go-pe/files/7169673/badpe.zip).
MD5 List:

1. 0be7c747a399d804caffa2584a7c9a21
2. 588a7e2bce94d494e85016096b8564d2
3. 9e7c4a4387fce4b8c72863c14e7cf706

**Comment by scudette** [comment]
The bug was around the total number of import table entries limit which was not enforced in some cases. I added some more checks to make sure the import table is not too large but it appears to be junk for those samples. I checked pefile and it actually aborts parsing after some errors and ignore all the imports completely. Maybe this is a good approach?

We currently try to parse the data anyway but end up with garbage which might not be as useful.
